mcp23017_driver/mcp23017.h

```diff
@@ -0,0 +1,351 @@
+ * @date 16/11/2023
+ * @author Enrico Dalla Croce (Kalsifer-742)
+ * 
+ * @brief Driver to configure and operate the MCP23X17 I/O Expander
+*/
+
+/**
+ * @paragraph Documentation
+ * 
+ * This file contains declarations and partial documentation for the MCP23017 I/O expander driver.
+ * Specific functions are fully documented to serve as examples for similar functions,
+ * assuming a consistent behavior across the driver.
+ * Specific functions should be thoroughly documented, outlining any unique behaviors.
+ * For comprehensive details, consult the MCP23017 datasheet.
+*/
+
+///Include guards
+#ifndef mcp_23017_h
+#define mcp_23017_h
+
+/// Includes
+/// Import of I2C_HandleTypeDef and HAL_StatusTypeDef
+#include "main.h"
+/// Import additional types, like: uint8_t
+#include <stdint.h>
+/// Import bool type
+#include <stdbool.h>
+
+/**
+ * @struct MCP23017_t
+ * @brief rapresents the configuration register (IOCON)
+*/
+typedef struct {
+    /**
+     * @brief Controls how the registers are addressed
+     * 1 = The registers associated with each port are separated into different banks.
+     * 0 = The registers are in the same bank (addresses are sequential).
+     */
+    uint8_t bank,
+    /**
+     * @brief INT Pins Mirror bit
+     * 1 = The INT pins are internally connected
+     * 0 = The INT pins are not connected. INTA is associated with PORTA and INTB is associated with PORTB
+     */
+    mirror,
+    /**
+     * @brief Sequential Operation mode bit
+     * 1 = Sequential operation disabled, address pointer does not increment.
+     * 0 = Sequential operation enabled, address pointer increments
+     */
+    seqop,
+    /**
+     * @brief Slew Rate control bit for SDA output
+     * 1 = Slew rate disabled
+     * 0 = Slew rate enabled
+     */
+    disslw,
+    /**
+     * @brief Address pins are always enabled on the MCP23017. this setting is only for the MCP23S17.
+     */
+    haen_MCP23S17_only,
+    /**
+     * @brief Configures the INT pin as an open-drain output
+     * 1 = Open-drain output (overrides the INTPOL bit.)
+     * 0 = Active driver output (INTPOL bit sets the polarity.)
+     */
+    odr,
+    /**
+     * @brief This bit sets the polarity of the INT output pin
+     * 1 = Active-high
+     * 0 = Active-low
+     */
+    intpol;
+} MCP23017_configuration_t;
+
+MCP23017_configuration_t default_configuration = {
+    .bank   = 0,
+    .mirror = 0, 
+    .seqop  = 0,
+    .disslw = 0,
+    .odr    = 0,
+    .intpol = 0
+};
+
+/**
+ * @struct MCP23017_t
+ * @brief rapresents the MCP23017 device
+ */
+typedef struct {
+    I2C_HandleTypeDef *hi2c; ///I2C handle used for communication.
+    uint8_t device_address; ///I2C address of the MCP23017 device.
+    uint8_t i2c_timeout; ///Timeout value for I2C communication in milliseconds
+} MCP23017_t;
+
+/**
+ * @typedef
+ * @brief Specifies whether to refer to the PORTA or PORTB register.
+ */
+typedef enum {
+    port_a,
+    port_b,
+} port_t;
+
+typedef enum {
```

Comment:
Change all this typedef to defines and keep then only if necessary (logic_high and flag_high can be one define only for example)

Reply:
Removed all typedefs that only add verbosity